### ISSUE TITLE
What should happen when update() or commit() is called on a nested lock?

### ISSUE THREAD

**Issue by chrishtr** [WICG/display-locking#59]
If an element e is locked, and its ancestor is also locked, what should happen if we call e.update() or e.commit()?

Option 1: reject
Option 2: update/commit all ancestor locks (this is similar / the same as not allowing nested locks, and instead delegating to an ancestor lock if it exists)
Option 3: synchronously "unlock" the current lock for future lifecycles (which amounts to removing the lock and otherwise doing nothing, if in a nested lock

Use cases:
a. Want to unlock some content that may be under another lock, and then forget about it. Don't want to accidentally leave the content unlocked if the ancestor unlocks in the future.
b. Want to update some content in order to measure its layout.
c. Want to update some content in order to pre-render it to prepare for quick final display.

Use case (b) needs to also know whether the update() or commit() or updateAndCommit() actually finished, or whether it stayed locked, in order to avoid forced layouts when reading back layout.

for commit() and updateAndCommit(), I think option 3 makes the most sense, because it solves use case (a).

For update(), option 2 would solve the most use cases, e.g. (b). But option 1 is more performant because it avoids doing work that is much larger than the display locked subtree.

None of the options solve the problem of knowing whether it's safe to read back layout though.

**Comment by chrishtr** [comment]
Option 4: wait around until the ancestor lock is unlocked, and only then proceed and resolve update or commit promises.

I think option 4 is a bad idea, because it can cause performance problems due to promises never resolving. See crbug.com/972498.

**Comment by vmpstr** [comment]
Option 2 has some implementation quirks at least with the naive approach. Consider the following:

lock1
  lock2

- we call lock1.update() and it passes all but 1 of the phases, so it needs one more phase to do.
- now we call lock2.update(), and start a new frame
- lock2 updates just 1 phase, and runs out of budget and earlies out. lock1 processes all of the previous phases it did, with lock2 preventing some of the phases. This means that from lock1 perspective it actually finishes the right phases there. Then it does its last phase and enters the locked state.
- lock2 is now back in the "want updates but nobody is doing them" state, which is option 4.

Basically to properly implement this, we need lock1 to be aware that there is an updating lock2 somewhere in its subtree and artificially delay finishing the update until all such locks are finished. 

I think this is still implementable, but it requires some bookkeeping that might be a bit tricky. I'll see if I can play around with it and see if it's not too bad

**Comment by fergald** [comment]
Option 4 only causes performance problems if it leave long-lived lifecycle observers. The observers could be smarter and no observe locks are nested, instead they could observe their outer lock(s).

I'm a bit confused by Option 2 which says "update/commit all ancestor locks". Why commit? Can't we just update everything up the chain of locks?

**Comment by vmpstr** [comment]
We can call update on the ancestor chain, but the problem we run into is that the updated locks above the current one need to be kept in the "updating" phase until all of the descendants finish updating. If the ancestor finishes too soon, then it goes back to being locked. 

I've tried to prototype this, but there are some cases where this is pretty hard to do and have it perform well. For instance, if we have a lock that is currently in the updating phase somewhere in our subtree, and we call acquire on the (possibly distant) ancestor, that ancestor needs to know that there is a lock that is updating in the subtree and enter this forced state. However, I don't see how we can do that without walking the whole subtree on the hunt for these updating locks. Specifically, this would mean that _all_  acquires would have to do a full subtree walk. That is bad. 

The alternative is that the updating lock at each begin frame does an ancestor walk to ensure that all of its locked ancestors are still forced. This is probably doable since we would only do this for updating locks and we're walking up not down. This can likely be optimized when we have smarter detection of "we're within a locked subtree". That way, locks that are not in locked subtrees can just early out and we recover most of the performance.

**Comment by chrishtr** [comment]
For nested locks, I think `update()` should throw, and `commit() & `updateAndCommit()` should just work as they do already.

The reason is that trying to update ancestor locks in concert with the current one (a) has a number of complications in corner cases such as running out of budget, or if the ancestor unlocks or locks halfway thorough, and (b) has performance that is hard to understand for the developer.

If the developer wants to cause a nested update to occur (e.g to measure layout in a nested lock), they can unlock descendant locks and then call `update()` on the root ancestor. After done, they can re-lock the descendants if desired.
